akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1156,15 +1157,20 @@ private[stream] object Collect {
 
       @tailrec private def pushOne(): Unit =
         if (buffer.isEmpty) {
-          if (isClosed(in)) completeStage()
-          else if (!hasBeenPulled(in)) pull(in)
+          if (isClosed(in)) {
+            if (todo == 0) completeStage()
+          } else if (!hasBeenPulled(in)) pull(in)
         } else if (buffer.peek().elem == NotYetThere) {
           if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
         } else buffer.dequeue().elem match {
           case Success(elem) ⇒
             push(out, elem)
             if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
-          case Failure(ex) ⇒ pushOne()
+          case Failure(e) ⇒
+            decider(e) match {
```

Comment:
There's also `holderCompleted` above. Would probably make sense to have only one code path for the same thing.